Remove more complex ARRAY_SIZE macro

lucasdemarchi opened the pull request.

tridge commented:
thanks, i prefer the simpler one, but I can't remember now why the more complex one was added. Maybe we should find the PR that added it and see if there was some reason?

lucasdemarchi commented:
@tridge I was the one who added it a long time ago.  The reason for the more complex variant is that it catches bugs like:

```c++
static const char *_table[10];

int foo(const char table[10])
{
    int n = ARRAY_SIZE(table);
    ...
}

...

foo(_table);

```

or

```c++
char *table;

int foo()
{
    int n = ARRAY_SIZE(table);
}
```

The second case is an obvious bug, in which we are using a pointer. The first is more subtle, the reason being that in C/C++ arrays decay to a pointer type when passed to functions.  I think when I added this I fixed 1 bug, but I'm not sure.

`git log --author=lucas --grep ARRAY_SIZE` points to:
7ba6f92eb52f5c124595c1fbdb8aebf768bdac67 , with more explanation for the origin of the macro.
